lib/validator.js

```diff
@@ -0,0 +1,159 @@
+"use strict";
+
+(() => {
+  function error(source, token, current, str) {
+    const consume_position = token.index;
+    const maxTokens = 5;
+    const line =
+      !source[token.index].type === "eof" ? source[consume_position].line :
+      source.length > 1 ? source[consume_position - 1].line :
+      1;
+
+    const precedingLine = lastLine(
+      tokensToText(sliceTokens(-maxTokens), { precedes: true })
+    );
+
+    const procedingTokens = sliceTokens(maxTokens);
+    const procedingText = tokensToText(procedingTokens);
+    const procedingLine = procedingText.split("\n")[0];
+
+    const spaced = " ".repeat(precedingLine.length) + "^ " + str;
+    const context = precedingLine + procedingLine + "\n" + spaced;
+
+    const inside = current ? `, inside \`${current.partial ? "partial " : ""}${current.type} ${current.name}\`` : "";
+    return `Validation error at line ${line}${inside}:\n${context}`;
+
+    function sliceTokens(count) {
+      return count > 0 ?
+        source.slice(consume_position, consume_position + count) :
+        source.slice(Math.max(consume_position + count, 0), consume_position);
+    }
+
+    function tokensToText(inputs, { precedes } = {}) {
+      const text = inputs.map(t => t.trivia + t.value).join("");
+      const nextToken = source[consume_position];
+      if (nextToken.type === "eof") {
+        return text;
+      }
+      if (precedes) {
+        return text + nextToken.trivia;
+      }
+      return text.slice(nextToken.trivia.length);
+    }
+
+    function lastLine(text) {
+      const splitted = text.split("\n");
+      return splitted[splitted.length - 1];
+    }
+  }
+
+  function groupDefinitions(all) {
+    const unique = new Map();
+    const duplications = new Set();
```

Comment:
Nit: “duplicates”?

Reply:
Actually, it’s fine to leave it. Would mean changing this in way too many places.

Reply:
That's why IDEs have variable renaming feature 😁